File: LeetCode - Top Interview 150/104.py
# 104. Maximum Depth of Binary Tree
# https://leetcode.com/problems/maximum-depth-of-binary-tree/


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # BFS - Breadth First Search
    def maxDepth1(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        queue = []
        max_depth = 0

        queue.append((1, root))
        while queue:
            depth, current = queue.pop(0)
            max_depth = max(max_depth, depth)

            if current.left:
                queue.append((depth + 1, current.left))

            if current.right:
                queue.append((depth + 1, current.right))

        return max_depth

    # DFS - Depth First Search
    def maxDepth2(self, root: Optional[TreeNode]) -> int:
        def dfs(node, current_depth):
            if not node:
                return current_depth - 1

            return max(
                dfs(node.left, current_depth + 1), dfs(node.right, current_depth + 1)
            )

        return dfs(root, 1)

File: LeetCode - Algorithm/Binary Tree/104.java
// 104. Maximum Depth of Binary Tree
// https://leetcode.com/problems/maximum-depth-of-binary-tree/

import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int result = 0;

        Queue<TreeNode> queue = new LinkedList<>();

        queue.offer(root);

        while (!queue.isEmpty()) {
            int queueSize = queue.size();

            for (int i = 0; i < queueSize; i++) {
                TreeNode node = queue.poll();

                if (node.left != null) {
                    queue.offer(node.left);
                }

                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            result += 1;
        }

        return result;
    }
}

/*
 * In Java,
 * 
 * Set<String> hashSet = new HashSet<>();
 * 
 * Map<String, Integer> hashMap = new HashMap<>();
 * 
 * Stack<String> stack = new Stack<>();
 * 
 * Queue<String> queue = new ArrayDeque<>();
 * -> queue.offer();
 * -> queue.poll();
 * 
 * ArrayDeque<String> deque = new ArrayDeque<>();
 * -> deque.addFirst();
 * -> deque.addLast();
 * -> deque.removeFirst();
 * -> deque.removeLast();
 */